domain/tutoringModel/ga.py

```diff
@@ -0,0 +1,312 @@
+
+
+class Genetische_Algorithm:
+    """Genetische Algorithmus  sdfasdfs"""
+
+    def __init__(
+        self,
+        learning_style=None,
+        learning_elements=None,
+    ):
+        self.learning_elements = learning_elements
+        self.best_population = None
+        self.le_coordinate = None
+        self.n_generation = 100
+        self.population = None
+        self.le_size = None
+        self.mutate_rate = 0.3
+        self.cross_rate = 0.9
+        self.pop_size = 80
+
+    def create_random_population(self, dict_coordinates) -> None:
+        """function for the calculation of the score.
+        position a also Initialise some populations
+        with Clustering if this is possible.
+        param dict_coordinates"""
+        self.le_coordinate = np.array(
+            [dict_coordinates[key] for key in dict_coordinates]
+        )
+        self.le_coordinate.reshape((len(dict_coordinates), 4))
+
+        positions = np.arange(1, self.le_size)
+        self.population = np.vstack(
+            [np.random.permutation(positions) for _ in range(self.pop_size)]
+        )
+
+    def valide_population(self):
+        """Function to add validation:
+        First Learning Element is fixed
+        in the Learning path."""
+
+        new_pop = np.zeros((self.pop_size, self.le_size), dtype=int)
+        new_pop[:, 0] = 0
+
+        new_pop[:, 1 : self.le_size] = self.population.copy()
+        return new_pop
+
+    def get_lines_paths(self, new_pop):
+        """This function extract all the positions X,Y,Z,K.
+        in the space of the generated trajectories"""
+        line_x = np.empty_like(new_pop, dtype=np.float64)
+        line_y = np.empty_like(new_pop, dtype=np.float64)
+        line_z = np.empty_like(new_pop, dtype=np.float64)
+        line_k = np.empty_like(new_pop, dtype=np.float64)
+
+        for i, j in enumerate(new_pop):
+            le_coord = self.le_coordinate[j]
+            line_x[i, :] = le_coord[:, 0]
+            line_y[i, :] = le_coord[:, 1]
+            line_z[i, :] = le_coord[:, 2]
+            line_k[i, :] = le_coord[:, 3]
+        return line_x, line_y, line_z, line_k
+
+    def get_fitness(self, line_x, line_y, line_z, line_k):
+        """Funcion para Berechnung der Fitnessnote."""
+        # a score is evaluated for each individual, which is
+        # calculated using the fitness function: dtype=np.float64)
+        total_sum = (
+            np.square(np.diff(line_x))
+            + np.square(np.diff(line_y))
+            + np.square(np.diff(line_z))
+            + np.square(np.diff(line_k))
+        )
+        fitness = np.sqrt(np.sum(total_sum, 1))
+        return fitness
+
+    def crossover(self, parent, pop):
+        """function for the calculation
+        of crossovers between each individual."""
+        # two parents are randomly
+        # selected from the population, to pass on a part
+        # of their solution to their child.  """
+
+        if np.random.rand() < self.cross_rate:
```

Comment:
I removed it

Reply:
I probably was not precise enough. ^^ Instead of using np.random.rand() you should assign a new variable `rn = np.random.default_rng(seed)` since np.random.rand() is depcrecated. Do you know what I mean? If you have questions just ask me on discord.